lib/common/abstract-descriptor.h

```diff
@@ -0,0 +1,269 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_COMMON_ABSTRACT_DESCRIPTOR_H_
+#define FORTRAN_COMMON_ABSTRACT_DESCRIPTOR_H_
+
+#include "constexpr-bitset.h"
+
+namespace Fortran::common {
+
+// DescriptorInterface both serves to describe and validate the interface
+// that descriptor implementations must comply with in order to be used to
+// instantiate Fortran transformational intrinsic abstract implementation.
+// To validate the compliance of a descriptor implementation DESC with the
```

Comment:
If you used the CRTP paradigm instead, this validation might not be necessary.

Reply:
Right, I considered it and I did not like using CRTP only for interface purposes, but maybe it could also be a better place for transformational if I change them to be descriptor member functions in order to split the memory and logic aspects (and not expect to dynamic memory allocation for every transformational results as per your other comments).